README.md
# Animated flex

Flutter animation flex.

## example

<img src="./demo.gif" height="600">

### demo1
```
AnimatedFlex(
  children: [
    Container(
      color: Colors.red,
      height: 200,
      margin: EdgeInsets.only(top: 20),
    ),
    Container(
      color: Colors.green,
      height: 200,
      margin: EdgeInsets.only(top: 20),
    ),
    Container(
      color: Colors.blue,
      height: 200,
      margin: EdgeInsets.only(top: 20),
    ),
  ],
)
```
### demo2
```
AnimatedFlex(
  animationWidgetBuilder: (Animation animation, Widget child) {
    return Opacity(
      opacity: animation.value,
      child: Transform(
        transform: Matrix4.translationValues(
            20 * (1.0 - animation.value),
            20 * (1.0 - animation.value),
            0.0),
        child: child,
      ),
    );
  },
  children: [
    Container(
      color: Colors.red,
      height: 200,
      margin: EdgeInsets.only(top: 20),
    ),
    Container(
      color: Colors.green,
      height: 200,
      margin: EdgeInsets.only(top: 20),
    ),
    Container(
      color: Colors.blue,
      height: 200,
      margin: EdgeInsets.only(top: 20),
    ),
  ],
)
```
### demo3
```
Column(
  children: [
    SizedBox(
      height: 100,
      child: AnimatedFlex(
        animationWidgetBuilder: (Animation animation, Widget child) {
          return Opacity(
            opacity: animation.value,
            child: Transform(
              transform: Matrix4.translationValues(
                  20 * (1.0 - animation.value), 0, 0.0),
              child: child,
            ),
          );
        },
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        direction: Axis.horizontal,
        children: [
          Container(
            color: Colors.red,
            height: 100,
            width: 100,
          ),
          Container(
            color: Colors.green,
            height: 100,
            width: 100,
          ),
          Container(
            color: Colors.blue,
            height: 100,
            width: 100,
          ),
        ],
      ),
    ),
    SizedBox(
      height: 400,
      child: AnimatedFlex(
        delayed: Duration(milliseconds: 1000),
        animationWidgetBuilder: (Animation animation, Widget child) {
          return Opacity(
            opacity: animation.value,
            child: Transform(
              transform: Matrix4.translationValues(
                  20 * (1.0 - animation.value),
                  20 * (1.0 - animation.value),
                  0.0),
              child: child,
            ),
          );
        },
        children: [
          Container(
            color: Colors.red,
            height: 100,
            margin: EdgeInsets.only(top: 20),
          ),
          Container(
            color: Colors.green,
            height: 100,
            margin: EdgeInsets.only(top: 20),
          ),
          Expanded(
            child: Container(
              color: Colors.blue,
              // height: 100,
              margin: EdgeInsets.only(top: 20),
            ),
          ),
        ],
      ),
    ),
  ],
)
```
### demo4
```
AnimatedFlex(
  animationWidgetBuilder: (Animation animation, Widget child) {
    return Opacity(
      opacity: animation.value,
      child: Transform.scale(
          scale: 0.8 + animation.value * .2,
          child: child),
    );
  },
  children: [
    Container(
      color: Colors.red,
      height: 150,
      margin: EdgeInsets.only(top: 20),
    ),
    Container(
      color: Colors.green,
      height: 150,
      margin: EdgeInsets.only(top: 20),
    ),
    Container(
      color: Colors.blue,
      height: 150,
      margin: EdgeInsets.only(top: 20),
    ),
  ],
)
```

CHANGELOG.md
## [0.0.1] - 2020/9/27

- Custom animation effect flex



test/animation_flex_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:animation_flex/animation_flex.dart';

void main() {

}


lib/animated_flex.dart
library animated_flex;

import 'package:flutter/material.dart';

typedef AnimationWidgetBuilder = Widget Function(
    Animation animation, Widget child);

Widget _defaultAnimationWidgetBuilder(Animation animation, Widget child) {
  return Opacity(
    opacity: animation.value,
    child: Transform(
      transform:
          Matrix4.translationValues(0, 10 * (1.0 - animation.value), 0.0),
      child: child,
    ),
  );
}

///Custom animation effect flex
class AnimatedFlex extends StatefulWidget {
  final Axis direction;

  final MainAxisAlignment mainAxisAlignment;

  final MainAxisSize mainAxisSize;

  final CrossAxisAlignment crossAxisAlignment;

  final TextDirection textDirection;

  final VerticalDirection verticalDirection;

  final TextBaseline textBaseline;

  final Clip clipBehavior;

  final List<Widget> children;

  /// Animation duration
  final Duration duration;

  /// Delay animation execution
  final Duration delayed;

  /// Animation end event
  final VoidCallback onEnd;

  /// Custom animation effect
  final AnimationWidgetBuilder animationWidgetBuilder;

  const AnimatedFlex(
      {Key key,
      this.direction = Axis.vertical,
      this.mainAxisAlignment = MainAxisAlignment.start,
      this.mainAxisSize = MainAxisSize.max,
      this.crossAxisAlignment = CrossAxisAlignment.center,
      this.textDirection,
      this.verticalDirection = VerticalDirection.down,
      this.textBaseline,
      this.clipBehavior = Clip.hardEdge,
      this.children = const <Widget>[],
      this.duration = const Duration(milliseconds: 1000),
      this.delayed = const Duration(milliseconds: 0),
      this.animationWidgetBuilder = _defaultAnimationWidgetBuilder,
      this.onEnd})
      : super(key: key);

  @override
  _AnimatedFlexState createState() => _AnimatedFlexState();
}

class _AnimatedFlexState extends State<AnimatedFlex>
    with SingleTickerProviderStateMixin {
  AnimationController animationController;

  Axis get direction => widget.direction;

  MainAxisAlignment get mainAxisAlignment => widget.mainAxisAlignment;

  MainAxisSize get mainAxisSize => widget.mainAxisSize;

  CrossAxisAlignment get crossAxisAlignment => widget.crossAxisAlignment;

  TextDirection get textDirection => widget.textDirection;

  VerticalDirection get verticalDirection => widget.verticalDirection;

  TextBaseline get textBaseline => widget.textBaseline;

  Clip get clipBehavior => widget.clipBehavior;

  List<Widget> get children => widget.children;

  Duration get duration => widget.duration;

  Duration get delayed => widget.delayed;

  VoidCallback get onEnd => widget.onEnd;

  AnimationWidgetBuilder get animationWidgetBuilder =>
      widget.animationWidgetBuilder;

  @override
  void initState() {
    super.initState();
    animationController = AnimationController(vsync: this, duration: duration);
    Future.delayed(delayed, () {
      if(mounted) {
        animationController?.forward();
      }
    });
    animationController.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        onEnd?.call();
      }
    });
  }

  List<Widget> initAnimationList() {
    List<Widget> list = [];
    int count = children.length;
    for (int i = 0; i < count; i++) {
      final Animation<double> animation =
          Tween<double>(begin: 0.0, end: 1.0).animate(
        CurvedAnimation(
          parent: animationController,
          curve: Interval((1 / count) * i, 1.0, curve: Curves.fastOutSlowIn),
        ),
      );
      Widget child = children[i];
      if(child is Expanded) {
        Expanded newChild = child as Expanded;
        child = Expanded(child: animationWidgetBuilder(animation, newChild.child));
      } else {
        child = animationWidgetBuilder(animation, child);
      }
      list.add(child);
    }
    return list;
  }

  @override
  void dispose() {
    animationController?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
        animation: animationController,
        builder: (BuildContext context, Widget child) {
          return Flex(
            direction: direction,
            children: initAnimationList(),
            mainAxisAlignment: mainAxisAlignment,
            mainAxisSize: mainAxisSize,
            crossAxisAlignment: crossAxisAlignment,
            textDirection: textDirection,
            verticalDirection: verticalDirection,
            textBaseline: textBaseline,
            clipBehavior: clipBehavior,
          );
        });
  }
}


lib/animation_flex.dart
library animation_flex;

import 'package:flutter/material.dart';


typedef AnimationWidgetBuilder = Widget Function(
    Animation animation, Widget child);

Widget _defaultAnimationWidgetBuilder(Animation animation, Widget child) {
  return Opacity(
    opacity: animation.value,
    child: Transform(
      transform:
      Matrix4.translationValues(0, 10 * (1.0 - animation.value), 0.0),
      child: child,
    ),
  );
}

///Custom animation effect flex
class AnimationFlex extends StatefulWidget {
  final Axis direction;

  final MainAxisAlignment mainAxisAlignment;

  final MainAxisSize mainAxisSize;

  final CrossAxisAlignment crossAxisAlignment;

  final TextDirection textDirection;

  final VerticalDirection verticalDirection;

  final TextBaseline textBaseline;

  final Clip clipBehavior;

  final List<Widget> children;

  final Duration duration;

  /// Animation end event
  final VoidCallback onEnd;

  /// Custom animation effect
  final AnimationWidgetBuilder animationWidgetBuilder;

  const AnimationFlex(
      {Key key,
        this.direction = Axis.vertical,
        this.mainAxisAlignment = MainAxisAlignment.start,
        this.mainAxisSize = MainAxisSize.max,
        this.crossAxisAlignment = CrossAxisAlignment.center,
        this.textDirection,
        this.verticalDirection = VerticalDirection.down,
        this.textBaseline,
        this.clipBehavior = Clip.hardEdge,
        this.children = const <Widget>[],
        this.duration = const Duration(milliseconds: 1000),
        this.animationWidgetBuilder = _defaultAnimationWidgetBuilder,
        this.onEnd})
      : super(key: key);

  @override
  _AnimationFlexState createState() => _AnimationFlexState();
}

class _AnimationFlexState extends State<AnimationFlex>
    with SingleTickerProviderStateMixin {
  AnimationController animationController;

  Axis get direction => widget.direction;

  MainAxisAlignment get mainAxisAlignment => widget.mainAxisAlignment;

  MainAxisSize get mainAxisSize => widget.mainAxisSize;

  CrossAxisAlignment get crossAxisAlignment => widget.crossAxisAlignment;

  TextDirection get textDirection => widget.textDirection;

  VerticalDirection get verticalDirection => widget.verticalDirection;

  TextBaseline get textBaseline => widget.textBaseline;

  Clip get clipBehavior => widget.clipBehavior;

  List<Widget> get children => widget.children;

  Duration get duration => widget.duration;

  VoidCallback get onEnd => widget.onEnd;

  AnimationWidgetBuilder get animationWidgetBuilder => widget.animationWidgetBuilder;

  @override
  void initState() {
    super.initState();
    animationController = AnimationController(vsync: this, duration: duration);
    animationController.forward();
    animationController.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        onEnd?.call();
      }
    });
  }

  List<Widget> initAnimationList() {
    List<Widget> list = [];
    int count = children.length;
    for (int i = 0; i < count; i++) {
      final Animation<double> animation =
      Tween<double>(begin: 0.0, end: 1.0).animate(
        CurvedAnimation(
          parent: animationController,
          curve: Interval((1 / count) * i, 1.0, curve: Curves.fastOutSlowIn),
        ),
      );
      list.add(animationWidgetBuilder(animation, children[i]));
    }
    return list;
  }

  @override
  void dispose() {
    animationController?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
        animation: animationController,
        builder: (BuildContext context, Widget child) {
          return Flex(
            direction: direction,
            children: initAnimationList(),
            mainAxisAlignment: mainAxisAlignment,
            mainAxisSize: mainAxisSize,
            crossAxisAlignment: crossAxisAlignment,
            textDirection: textDirection,
            verticalDirection: verticalDirection,
            textBaseline: textBaseline,
            clipBehavior: clipBehavior,
          );
        });
  }
}


example/lib/main.dart
import 'package:animated_flex/animated_flex.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Widget> list = [Demo1(), Demo2(), Demo3(), Demo4()];

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: list.length,
      child: Scaffold(
        appBar: AppBar(
          title: Text('Animated flex'),
          bottom: TabBar(
            tabs: [
              const Text('Demo1'),
              const Text('Demo2'),
              const Text('Demo3'),
              const Text('Demo4'),
            ],
          ),
        ),
        body: TabBarView(
          children: [for (Widget widget in list) widget],
        ),
      ),
    );
  }
}

class Demo1 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Padding(
        padding: EdgeInsets.all(20),
        child: AnimatedFlex(
          children: [
            Container(
              color: Colors.red,
              height: 200,
              margin: EdgeInsets.only(top: 20),
            ),
            Container(
              color: Colors.green,
              height: 200,
              margin: EdgeInsets.only(top: 20),
            ),
            Container(
              color: Colors.blue,
              height: 200,
              margin: EdgeInsets.only(top: 20),
            ),
          ],
        ),
      ),
    );
  }
}

class Demo2 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Padding(
        padding: EdgeInsets.all(20),
        child: AnimatedFlex(
          animationWidgetBuilder: (Animation animation, Widget child) {
            return Opacity(
              opacity: animation.value,
              child: Transform(
                transform: Matrix4.translationValues(
                    20 * (1.0 - animation.value),
                    20 * (1.0 - animation.value),
                    0.0),
                child: child,
              ),
            );
          },
          children: [
            Container(
              color: Colors.red,
              height: 200,
              margin: EdgeInsets.only(top: 20),
            ),
            Container(
              color: Colors.green,
              height: 200,
              margin: EdgeInsets.only(top: 20),
            ),
            Container(
              color: Colors.blue,
              height: 200,
              margin: EdgeInsets.only(top: 20),
            ),
          ],
        ),
      ),
    );
  }
}

class Demo3 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Padding(
        padding: EdgeInsets.all(20),
        child: Column(
          children: [
            SizedBox(
              height: 100,
              child: AnimatedFlex(
                animationWidgetBuilder: (Animation animation, Widget child) {
                  return Opacity(
                    opacity: animation.value,
                    child: Transform(
                      transform: Matrix4.translationValues(
                          20 * (1.0 - animation.value), 0, 0.0),
                      child: child,
                    ),
                  );
                },
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                direction: Axis.horizontal,
                children: [
                  Container(
                    color: Colors.red,
                    height: 100,
                    width: 100,
                  ),
                  Container(
                    color: Colors.green,
                    height: 100,
                    width: 100,
                  ),
                  Container(
                    color: Colors.blue,
                    height: 100,
                    width: 100,
                  ),
                ],
              ),
            ),
            SizedBox(
              height: 400,
              child: AnimatedFlex(
                delayed: Duration(milliseconds: 1000),
                animationWidgetBuilder: (Animation animation, Widget child) {
                  return Opacity(
                    opacity: animation.value,
                    child: Transform(
                      transform: Matrix4.translationValues(
                          20 * (1.0 - animation.value),
                          20 * (1.0 - animation.value),
                          0.0),
                      child: child,
                    ),
                  );
                },
                children: [
                  Container(
                    color: Colors.red,
                    height: 100,
                    margin: EdgeInsets.only(top: 20),
                  ),
                  Container(
                    color: Colors.green,
                    height: 100,
                    margin: EdgeInsets.only(top: 20),
                  ),
                  Expanded(
                    child: Container(
                      color: Colors.blue,
                      // height: 100,
                      margin: EdgeInsets.only(top: 20),
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class Demo4 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Padding(
        padding: EdgeInsets.all(20),
        child: AnimatedFlex(
          animationWidgetBuilder: (Animation animation, Widget child) {
            return Opacity(
              opacity: animation.value,
              child: Transform.scale(
                  scale: 0.8 + animation.value * .2,
                  child: child),
            );
          },
          children: [
            Container(
              color: Colors.red,
              height: 150,
              margin: EdgeInsets.only(top: 20),
            ),
            Container(
              color: Colors.green,
              height: 150,
              margin: EdgeInsets.only(top: 20),
            ),
            Container(
              color: Colors.blue,
              height: 150,
              margin: EdgeInsets.only(top: 20),
            ),
          ],
        ),
      ),
    );
  }
}
